pkg/cri/server/container_log_reopen.go

```diff
@@ -35,18 +35,5 @@ func (c *criService) ReopenContainerLog(ctx context.Context, r *runtime.ReopenCo
 	if container.Status.Get().State() != runtime.ContainerState_CONTAINER_RUNNING {
 		return nil, errors.New("container is not running")
 	}
-
-	// Create new container logger and replace the existing ones.
```

Comment:
😂 The kubelet use reopen api to rotate the log. If you use log plugin without involving shim, I think this idea might break existing usage.

It might be other topic: Kubelet can't prevent the container from huge log outputing. And there is one pull request using throttling method to prevent the huge outputting and disk pressure.

It is related to this change. I think we should file KEP to kubernetes community to enhance the log management.

ref: https://github.com/containerd/containerd/pull/5055

Reply:
I see log-uri=file  doesn't support to  rotate the log,
I don't want containerd(cri) to  rotate the log,it wll use a lot of memory.
maybe  log-uri=binary is suitable. use log-uri=binary we can  not only support json-file, flutend,syslog  but also support to rotate the log .

Reply:
currently, the ecosystem works with rotate and it is managed by kubelet. And kubelet is controling that log-path.

> I don't want containerd(cri) to rotate the log,it wll use a lot of memory.

I don't think it should be accepted. No rotate means that the disk is at risk. It is easy to run into the node disk pressure.

The enhancement is always welcome. But it should not break existing functionality.
That is why I said that we should enhance the kubernetes container log manager instead of containerd side.

Reply:
how about this  enhancement ?
add some config like 
```
[plugins."io.containerd.grpc.v1.cri".log_driver]
  [plugins."io.containerd.grpc.v1.cri".log_driver.json]
	path = "json-driver"
```
the ps tree will  like this which  refter to nerdctl
```
root      25872  1.0  0.0 711736 12184 ?        Sl   12:27   0:00 /usr/bin/containerd-shim-runc-v2 -namespace default -id e3dca6cab440db9c117520c3e8cdc197143e5816799b24c7e318216eefea3c
root      25882  3.5  0.1 729956 19328 ?        Sl   12:27   0:00  \_ /usr/local/bin/json-driver logpath
root      25899  0.9  0.0   4752   964 ?        Ss   12:27   0:00  \_ top
```

Reply:
Would you please describe the detail in the issue? Including the impact and compatibility. Thanks

Reply:
default  log-uri=fifo will compatible with previous versions
If use log-uri=file ,config like this, we can use logrotate to  rotate container log, instead of containerd,if os support  logrotate
```
[plugins."io.containerd.grpc.v1.cri"]
  max_container_log_line_size = 16384
  enable_log_file = true
  enable_log_rotate = true
  enable_log_binary = false
```
if  enable_log_binary = true ,will use log-uri=binary(some os may not support  logrotate),config like this
```
[plugins."io.containerd.grpc.v1.cri".log_driver]
  [plugins."io.containerd.grpc.v1.cri".log_driver.json]
    path = "json-driver"
    agrs = ""
  [plugins."io.containerd.grpc.v1.cri".log_driver.syslog]
    path = "syslog-driver"
    agrs = ""
   [plugins."io.containerd.grpc.v1.cri".log_driver.fluentd]
   path = "fluentd-driver"
   agrs = ""
```
fluentd-driver、  json-driver and syslog-driver  can be implemented  together or seperated
nerdctl support  fluentd-driver、  json-driver and syslog-driver
https://github.com/containerd/nerdctl/blob/5aa29d2ee6645285f54306ec6f5dbd79a8216bb3/pkg/logging/logging.go

Reply:
I see kubelet  code  rotateLogs() ->  rotateLog -> c.rotateLatestLog -> ReopenContainerLog ,I'm not familiar with kubelet,
maybe   enhance ReopenContainerLog()  is appropriate

Reply:
add logtheme to   distinguish fifo and file theme,It is compatible with kuberntes.